### PR TITLE
[Merged by Bors] - feat: `lake exe mk_all` as a Lean executable

### PR DESCRIPTION
**Pull request by adomani**
Running
```bash
lake exe mk_all --git
```
should have the same effect as running
```bash
./script/mk_all.sh
```

that is, it creates the files `Mathlib.lean`, `Mathlib/Tactic.lean`, `Archive.lean`, `Counterexamples.lean`.

It does *not* create analogous files for `Cache` and `LongestPole`.

```bash
lake exe mk_all
```
is similar, but uses all the `.lean` files in `Mathlib`, not just the Git-managed ones.

See #11874 for using the script in CI.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by YaelDillies** [comment]
@joneugster, do you think the same thing here as on #8361?

**Comment by joneugster** [comment]
Not necessarily. I personally care about import-analysis tools being outside of mathlib because I want to use them in my projects without importing mathlib.

But generally, while having separate packages with useful utilities is nice for projects not depending on mathlib, with the current in-development state of lake, it's still quite tedious to keep all packages in sync and up to date...

And finally, it's quite quick to extract useful stuff out of mathlib, once somebody concretely needs it. (The only reason I see to directly PR to a separate repo is that one can save some reviewer capacity if it's only utility-code)

**Comment by adomani** [comment]
I also think that "import" requirements are quite project-specific.

Even this tool is not just "import all files and/or all libraries":
* `Cache` is not created;
* `Mathlib/Tactic` should also be created.

`Std` has its own tool that does something way more sophisticated.

It seems like one of these commands that are hard to implement in a way that pleases most projects.

**Comment by grunweg** [comment]
I just reviewed the script (and some CI bits) in #11874. Summarising what I said there: looks very good, I only have minor comments.
awaiting-review

File: scripts/mk_all.lean
Comment (grunweg) [inline scripts/mk_all.lean:94]: Asking for me to learn: why are you quoting `Cli` and not `Cmd`, i.e. why does this line *not* look like this?
`def mkAll : Cmd := `[Cmd|`
That line doesn't work, this one does - if you'd like to tell me, that is welcome. It need not block this review at all.

Comment (adomani) [reply scripts/mk_all.lean:94]: This is completely unsubstantiated speculation: `Cli` makes me think of... well the command-line, while `Cmd` makes me think of a Lean `command`.  Maybe this is the distinction?  So, in my mind, this reads as "I'm defining a `Cmd` (in Lean), by using `Cli` syntax on the command-line".

Comment (grunweg) [reply scripts/mk_all.lean:94]: No matter if this is true, it's certainly a useful mnemonic :-)

Comment (Vierkantor) [reply scripts/mk_all.lean:94]: I believe the `Cli` here refers to the [`lean4-cli`](https://github.com/leanprover/lean4-cli) package.

File: scripts/mk_all.lean
Comment (Vierkantor) [inline scripts/mk_all.lean:29]: Would it be worth to split this up into a function that returns a `List String` of the form `["Mathlib.Algebra.Algebra.Basic", "Mathlib.Algebra.Algebra.Defs", ...]`? I can imagine that form of output is more useful for other scripts.

File: scripts/mk_all.lean
Comment (Vierkantor) [inline scripts/mk_all.lean:59]: I suppose this is okay for now. Probably we would want some sort of mechanism in Lake allowing us to differentiate build dependencies and dev dependencies.

**Review by Vierkantor — APPROVED** [review]
This looks plausible enough to me but I don't have any real Lean scripting experience so I'm hesitant to give definitive judgement.

File: scripts/mk_all.lean
Comment (grunweg) [reply scripts/mk_all.lean:94]: :bulb: Now it makes sense! Thanks a lot!

File: scripts/mk_all.lean
Comment (grunweg) [reply scripts/mk_all.lean:29]: Case in point: I need this functionality in #13245. (An array of strings is also fine for me.)

To share it, presumably this ought to go in a better place than this script (in `Mathlib/Util` perhaps? or in `Lake`).
In any case, moving this to a better place can also happen in a follow-up.

Comment (adomani) [reply scripts/mk_all.lean]: Thanks, I like this suggestions and I implemented it!

File: scripts/mk_all.lean
Comment (adomani) [reply scripts/mk_all.lean:63]: Good point!  I added a `TODO` in the doc-module highlighting this point.

**Review by Vierkantor — APPROVED** [review]
Thanks for the changes! I'm going to hold off on merging as I said, so let's kick this back on the maintainer queue:

maintainer merge

**Comment by grunweg** [comment]
I've proposed moving the common code into `Mathlib/Utils` in #13339.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**
